### PR TITLE
LTD-306: Flagging and routing rules changes

### PR DESCRIPTION
**Pull request by saruniitr**
## Change description

Multiple clc entries, destinations, case types can be added as comma separated values. Similarly multiple CLC entry groups can also be added as comma separated values.

Eg,
Multiple individual entries: `ML1a,ML22b,ML8a21`
Multiple groups: `ML8,ML1,ML22`

As the flags are involved the routing rules changes are also included in this ticket. Routing rules can now contain cases "that contain" certain flags and "that doesn't contain" certain flags. User can add multiple such conditions and they are saved in `flags_to_include` and `flags_to_exclude` fields.

### CODE REVIEW

File: api/flags/serializers.py [hunk 1]
```diff
@@ -27,6 +27,17 @@ class FlagReadOnlySerializer(serializers.Serializer):
...
 
 
+class FlaggingRuleReadOnlySerializer(serializers.Serializer):
+    level = serializers.CharField(read_only=True)
+    status = serializers.CharField(read_only=True)
+    matching_values = serializers.ListField()
+    matching_groups = serializers.ListField()
+
+
+class FlagwithFlaggingRulesReadOnlySerializer(FlagReadOnlySerializer):
```
Comment (saruniitr) [inline api/flags/serializers.py:37]: When adding flags to the routing rules we need to display flagging rules in the right side menu hence it is added.

File: api/flags/serializers.py [hunk 2]
```diff
@@ -140,20 +148,26 @@ class Meta:
             "level",
             "flag",
             "status",
-            "matching_value",
+            "matching_values",
+            "matching_groups",
             "is_for_verified_goods_only",
         )
-        validators = [
```
Comment (saruniitr) [inline api/flags/serializers.py:146]: User only need to enter `matchinv_values` or `matching_groups` hence we cannot use them to determine if this is a duplicate rule.

File: api/workflow/routing_rules/models.py [hunk 3]
```diff
@@ -29,7 +29,8 @@ class RoutingRule(TimestampableModel):
     # optional fields that are required depending on values in additional_rules
     user = models.ForeignKey(GovUser, related_name="routing_rules", on_delete=models.DO_NOTHING, blank=True, null=True)
     case_types = models.ManyToManyField(CaseType, related_name="routing_rules", blank=True)
-    flags = models.ManyToManyField(Flag, related_name="routing_rules", blank=True)
+    flags_to_include = models.ManyToManyField(Flag, related_name="routing_rules", blank=True)
+    flags_to_exclude = models.ManyToManyField(Flag, related_name="exclude_routing_rules", blank=True)
```
Comment (saruniitr) [inline api/workflow/routing_rules/models.py:33]: A routing rule can now be applied to cases that contain certain flags and doesn't contain some other flags, these are used to save those flags.

File: api/workflow/routing_rules/serializers.py [hunk 4]
```diff
@@ -96,41 +104,46 @@ def __init__(self, *args, **kwargs):
...
-                self.initial_data["flags"] = []
+        return validated_data
 
 
 # flattened serializer for editing purposes
 class SmallRoutingRuleSerializer(serializers.ModelSerializer):
     class Meta:
         model = RoutingRule
         fields = "__all__"
+        read_only_fields = (
```
Comment (saruniitr) [inline api/workflow/routing_rules/serializers.py:136]: this serializer is used when retrieving the details during editing the rule, hence it is read only.

File: runtime.txt [hunk 5]
```diff
@@ -1 +1 @@
-python-3.7.7
+python-3.7.9
```
Comment (saruniitr) [inline runtime.txt:1]: Latest master changes are missing in this branch hence this is added temporarily.

**Review by richtier — APPROVED** [review]

**Pull request merged**